File: src/com/demonwav/mcdev/creator/BukkitProjectSettingsWizard.java
package com.demonwav.mcdev.creator;

import com.demonwav.mcdev.exception.MinecraftSetupException;
import com.demonwav.mcdev.platform.bukkit.BukkitProjectConfiguration;

import com.intellij.ide.util.projectWizard.ModuleWizardStep;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.ui.MessageType;
import com.intellij.openapi.ui.popup.Balloon;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.ui.awt.RelativePoint;
import org.apache.commons.lang.WordUtils;
import org.jetbrains.annotations.NotNull;

import javax.swing.JComboBox;
import javax.swing.JComponent;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTextField;

public class BukkitProjectSettingsWizard extends ModuleWizardStep {

    private static final String pattern = "(\\s*(\\w+)\\s*(,\\s*\\w+\\s*)*,?|\\[?\\s*(\\w+)\\s*(,\\s*\\w+\\s*)*])?";

    private JPanel panel;
    private JTextField pluginNameField;
    private JTextField pluginVersionField;
    private JTextField mainClassField;
    private JTextField descriptionField;
    private JTextField authorsField;
    @Deprecated private JTextField additionAuthorsField;
    private JTextField websiteField;
    private JTextField prefixField;
    private JComboBox loadBox;
    private JTextField loadBeforeField;
    private JTextField dependField;
    private JTextField softDependField;
    private JLabel title;

    private BukkitProjectConfiguration settings = new BukkitProjectConfiguration();
    private MinecraftProjectCreator creator;

    public BukkitProjectSettingsWizard(@NotNull MinecraftProjectCreator creator) {
        super();
        this.creator = creator;
    }

    @Override
    public JComponent getComponent() {
        switch (creator.getType()) {
            case BUKKIT:
                title.setText("<html><font size=\"5\">Bukkit Settings</font></html>");
                break;
            case SPIGOT:
                title.setText("<html><font size=\"5\">Spigot Settings</font></html>");
                break;
            case PAPER:
                title.setText("<html><font size=\"5\">Paper Settings</font></html>");
                break;
        }

        pluginNameField.setText(WordUtils.capitalizeFully(creator.getArtifactId()));
        pluginVersionField.setText(creator.getVersion());
        mainClassField.setText(this.creator.getGroupId() + '.' + this.creator.getArtifactId()
                + '.' + WordUtils.capitalizeFully(this.creator.getArtifactId()));

        return panel;
    }

    @Override
    public boolean validate() throws ConfigurationException {
        try {
            if (pluginNameField.getText().trim().isEmpty()) {
                throw new MinecraftSetupException("empty", pluginNameField);
            }

            if (pluginVersionField.getText().trim().isEmpty()) {
                throw new MinecraftSetupException("empty", pluginVersionField);
            }

            if (mainClassField.getText().trim().isEmpty()) {
                throw new MinecraftSetupException("empty", mainClassField);
            }

            if (!additionAuthorsField.getText().matches(pattern)) {
                throw new MinecraftSetupException("bad", additionAuthorsField);
            }

            if (!loadBeforeField.getText().matches(pattern)) {
                throw new MinecraftSetupException("bad", loadBeforeField);
            }

            if (!dependField.getText().matches(pattern)) {
                throw new MinecraftSetupException("bad", dependField);
            }

            if (!softDependField.getText().matches(pattern)) {
                throw new MinecraftSetupException("bad", softDependField);
            }
        } catch (MinecraftSetupException e) {
            String message = e.getError();
            JBPopupFactory.getInstance().createHtmlTextBalloonBuilder(message, MessageType.ERROR, null)
                    .setFadeoutTime(4000)
                    .createBalloon()
                    .show(RelativePoint.getSouthWestOf(e.getJ()), Balloon.Position.below);
            return false;
        }
        return true;
    }

    @Override
    public void onStepLeaving() {
        super.onStepLeaving();
        settings.pluginName = pluginNameField.getText();
        settings.pluginVersion = pluginVersionField.getText();
        settings.mainClass = mainClassField.getText();
        settings.description = descriptionField.getText();
        this.settings.setAuthors(this.authorsField.getText());
        settings.website = websiteField.getText();
        settings.prefix = prefixField.getText();
        settings.load = loadBox.getSelectedIndex() == 0 ? BukkitProjectConfiguration.Load.POSTWORLD : BukkitProjectConfiguration.Load.STARTUP;
        this.settings.setLoadBefore(this.loadBeforeField.getText());
        this.settings.setDependencies(this.dependField.getText());
        this.settings.setSoftDependencies(this.softDependField.getText());
        creator.setSettings(settings);
    }

    @Override
    public void updateDataModel() {}
}

File: src/main/kotlin/com/demonwav/mcdev/creator/BukkitProjectSettingsWizard.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

@file:Suppress("Duplicates")

package com.demonwav.mcdev.creator

import com.demonwav.mcdev.asset.PlatformAssets
import com.demonwav.mcdev.platform.PlatformType
import com.demonwav.mcdev.platform.ProjectConfiguration
import com.demonwav.mcdev.platform.bukkit.BukkitProjectConfiguration
import com.demonwav.mcdev.platform.bukkit.data.LoadOrder
import com.demonwav.mcdev.util.firstOfType
import javax.swing.JComboBox
import javax.swing.JComponent
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JTextField
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.swing.Swing
import kotlinx.coroutines.withContext
import org.apache.commons.lang.WordUtils

class BukkitProjectSettingsWizard(private val creator: MinecraftProjectCreator) : MinecraftModuleWizardStep() {

    private lateinit var pluginNameField: JTextField
    private lateinit var pluginVersionField: JTextField
    private lateinit var mainClassField: JTextField
    private lateinit var panel: JPanel
    private lateinit var descriptionField: JTextField
    private lateinit var authorsField: JTextField
    private lateinit var websiteField: JTextField
    private lateinit var prefixField: JTextField
    private lateinit var loadOrderBox: JComboBox<*>
    private lateinit var loadBeforeField: JTextField
    private lateinit var dependField: JTextField
    private lateinit var softDependField: JTextField
    private lateinit var title: JLabel
    private lateinit var minecraftVersionBox: JComboBox<String>
    private lateinit var errorLabel: JLabel

    private var config: BukkitProjectConfiguration? = null

    override fun getComponent(): JComponent {
        return panel
    }

    override fun isStepVisible(): Boolean {
        return creator.configs.any { it is BukkitProjectConfiguration }
    }

    override fun updateStep() {
        config = creator.configs.firstOfType()
        if (config == null) {
            return
        }

        val buildSystem = creator.buildSystem ?: return

        val name = WordUtils.capitalize(buildSystem.artifactId.replace('-', ' '))
        pluginNameField.text = name
        pluginVersionField.text = buildSystem.version

        val conf = config ?: return

        if (creator.configs.indexOf(conf) != 0) {
            pluginNameField.isEditable = false
            pluginVersionField.isEditable = false
        }

        mainClassField.text = buildSystem.groupId.replace("-", "").toLowerCase() + "." +
            buildSystem.artifactId.replace("-", "").toLowerCase() + "." + name.replace(" ", "")

        if (creator.configs.size > 1) {
            mainClassField.text = mainClassField.text + conf.type.normalName
        }

        when (conf.type) {
            PlatformType.BUKKIT -> {
                title.icon = PlatformAssets.BUKKIT_ICON_2X
                title.text = "<html><font size=\"5\">Bukkit Settings</font></html>"
            }
            PlatformType.SPIGOT -> {
                title.icon = PlatformAssets.SPIGOT_ICON_2X
                title.text = "<html><font size=\"5\">Spigot Settings</font></html>"
            }
            PlatformType.PAPER -> {
                title.icon = PlatformAssets.PAPER_ICON_2X
                title.text = "<html><font size=\"5\">Paper Settings</font></html>"
            }
            else -> {
            }
        }

        CoroutineScope(Dispatchers.Swing).launch {
            try {
                withContext(Dispatchers.IO) { getVersionSelector(conf.type) }.set(minecraftVersionBox)
            } catch (e: Exception) {
                errorLabel.isVisible = true
            }
        }
    }

    override fun validate(): Boolean {
        return validate(
            pluginNameField,
            pluginVersionField,
            mainClassField,
            authorsField,
            dependField,
            pattern
        ) && minecraftVersionBox.selectedItem != null
    }

    override fun onStepLeaving() {
        val conf = config ?: return
        conf.data = BukkitProjectConfiguration.BukkitData(
            loadOrder = if (loadOrderBox.selectedIndex == 0) LoadOrder.POSTWORLD else LoadOrder.STARTUP,
            prefix = prefixField.text,
            minecraftVersion = minecraftVersionBox.selectedItem as? String ?: ""
        )
        conf.setLoadBefore(loadBeforeField.text)

        conf.base = ProjectConfiguration.BaseConfigs(
            pluginNameField.text,
            pluginVersionField.text,
            mainClassField.text,
            descriptionField.text,
            websiteField.text
        )

        conf.setAuthors(this.authorsField.text)
        conf.setDependencies(this.dependField.text)
        conf.setSoftDependencies(this.softDependField.text)
    }

    override fun updateDataModel() {}
}
